### PR TITLE
Initialize array with known size

### PR DESCRIPTION
**Pull request by JakubOnderka**
We known in advance the array size, so it makes sense to avoid reallocation when adding new elements

### CODE REVIEW

File: library.c [hunk 1]
```diff
@@ -3380,7 +3380,7 @@ redis_mbulk_reply_raw(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock, zval
         return FAILURE;
     }
     zval z_multi_result;
-    array_init(&z_multi_result); /* pre-allocate array for multi's results. */
+    array_init_size(&z_multi_result, numElems); /* pre-allocate array for multi's results. */
```
Comment (michael-grunder) [inline library.c]: I think we need to account for `numElems == -1`

File: library.c [hunk 2]
```diff
@@ -3409,7 +3409,7 @@ redis_mbulk_reply_double(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock, zv
         return FAILURE;
     }
 
-    array_init(&z_multi_result);
+    array_init_size(&z_multi_result, numElems);
```
Comment (michael-grunder) [inline library.c]: Same as the previous comment. `numElems` might be `-1`.

**Review by michael-grunder — CHANGES_REQUESTED** [review]
Good optimization, just need to account for `*-1\r\n`

**Comment by JakubOnderka** [comment]
> Good optimization, just need to account for `*-1\r\n`

Fixed. Now even when number of elements is -1 or 0, immutable empty array is used – this will avoid even more allocations.

File: redis.c [hunk 3]
```diff
@@ -2066,6 +2069,15 @@ redis_sock_read_multibulk_multi_reply_loop(INTERNAL_FUNCTION_PARAMETERS,
...
     fold_item *fi;
+    int count = 0;
+
+    /* Count issued commands so we can preallocate array for results */
+    for (fi = redis_sock->head; fi; /* void */) {
+        count++;
+        if (fi) fi = fi->next;
+    }
+
+    array_init_size(z_tab, count);
```
Comment (michael-grunder) [inline redis.c]: It would be interesting to profile whether this is actually faster in practice, given that the `HashTable` grows exponentially.

Comment (michael-grunder) [reply redis.c]: I think we should remove this commit from the PR until there are numbers showing it improves performance on average.

**Review by michael-grunder — CHANGES_REQUESTED** [review]
This is great. As in my comment I think we need numbers to justify double-iterating the multi/pipeline linked list, but if you want to remove that commit I'll rebase the other changes in this PR into develop.

**Comment by JakubOnderka** [comment]
Cherry picked to separate PR #2583 and removed from this branch, so it should be possible to merge it.

**Review by michael-grunder — APPROVED** [review]

**Review by yatsukhnenko — APPROVED** [review]

**Comment by michael-grunder** [comment]
Merged, thanks!

**Pull request merged**